compact_str/src/lib.rs

```diff
@@ -2392,4 +2427,88 @@ impl DoubleEndedIterator for Drain<'_> {
 
 impl FusedIterator for Drain<'_> {}
 
+/// A possible error value if allocating or resizing a [`CompactString`] failed.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct ReserveError(());
+
+impl fmt::Display for ReserveError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("Cannot allocated memory to hold CompactString")
+    }
+}
+
+#[cfg(feature = "std")]
+#[cfg_attr(docsrs, doc(cfg(feature = "std")))]
+impl std::error::Error for ReserveError {}
+
+/// A possible error value if [`ToCompactString::try_to_compact_string()`] failed.
+#[derive(Debug, Clone, Copy, PartialEq)]
+#[non_exhaustive]
+pub enum ToCompactStringError {
+    /// Cannot allocated memory to hold CompactString
+    Reserve(ReserveError),
+    /// [`Display::fmt()`][core::fmt::Display::fmt] returned an error
+    Fmt(fmt::Error),
+}
+
+impl fmt::Display for ToCompactStringError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            ToCompactStringError::Reserve(err) => err.fmt(f),
+            ToCompactStringError::Fmt(err) => err.fmt(f),
+        }
+    }
+}
+
+impl From<ReserveError> for ToCompactStringError {
+    #[inline]
+    fn from(value: ReserveError) -> Self {
+        Self::Reserve(value)
+    }
+}
+
+impl From<fmt::Error> for ToCompactStringError {
+    #[inline]
+    fn from(value: fmt::Error) -> Self {
+        Self::Fmt(value)
+    }
+}
+
+#[cfg(feature = "std")]
+#[cfg_attr(docsrs, doc(cfg(feature = "std")))]
+impl std::error::Error for ToCompactStringError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        match self {
+            ToCompactStringError::Reserve(err) => Some(err),
+            ToCompactStringError::Fmt(err) => Some(err),
+        }
+    }
+}
+
+trait UnwrapWithMsg {
+    type T;
+
+    fn unwrap_with_msg(self) -> Self::T;
+}
+
+impl<T, E: fmt::Display> UnwrapWithMsg for Result<T, E> {
```

Comment:
unwrap actually prints the error causing the panic https://doc.rust-lang.org/src/core/result.rs.html#1071-1073

Reply:
`unwrap_failed()` uses `{error:?}`: <https://doc.rust-lang.org/src/core/result.rs.html#1651-1653>, which is not very descriptive in most cases.

Reply:
With `.unwrap()` the message reads (in nightly):

> thread 'tests::test_alloc_excessively_long_string' panicked at src/tests.rs:1832:5:  
called `Result::unwrap()` on an `Err` value: ReserveError(())

With `.unwrap_with_msg()` the message reads:

> thread 'tests::test_alloc_excessively_long_string' panicked at src/tests.rs:1832:5:  
Cannot allocated memory to hold CompactString

Also the generated code is a tiny bit better, because the `error` is consumed by value, without a `&dyn` reference. This way no vtables or trampoline functions have to be generated.

Reply:
Thanks for explanation!